extensions/amp-access/0.1/amp-access-source.js

```diff
@@ -0,0 +1,621 @@
+   * @param {!Function} scheduleViewFn
+   * @param {!Element} accessElement
+   */
+  constructor(ampdoc, configJson, readerIdFn, scheduleViewFn, accessElement) {
+
+    /** @const */
+    this.ampdoc = ampdoc;
+
+    /** @const */
+    this.getReaderId_ = readerIdFn;
+
+    /** @const */
+    this.scheduleView_ = scheduleViewFn;
+
+    /** @const */
+    this.accessElement_ = accessElement;
+
+    /** @const @private {boolean} */
+    this.isServerEnabled_ = isExperimentOn(ampdoc.win, 'amp-access-server');
+
+    /** @const @private {boolean} */
+    this.isJwtEnabled_ = isExperimentOn(ampdoc.win, 'amp-access-jwt');
+
+    /** @const {!AccessType} */
+    this.type_ = this.buildConfigType_(configJson);
+
+    /** @const {!JsonObject} */
+    this.loginConfig_ = this.buildConfigLoginMap_(configJson);
+
+    /** @const {?JsonObject} */
+    this.authorizationFallbackResponse_ =
+        configJson['authorizationFallbackResponse'];
+
+    /** @const {?string} */
+    this.namespace_ = configJson['namespace'];
+
+    // TODO: Move type back
+    /** @const {!AccessTypeAdapterDef} */
+    this.adapter_ = this.createAdapter_(configJson);
+
+    /** @const @private {!string} */
+    this.pubOrigin_ = getSourceOrigin(ampdoc.win.location);
+
+    /** @const @private {!../../../src/service/vsync-impl.Vsync} */
+    this.vsync_ = Services.vsyncFor(ampdoc.win);
+
+    /** @const @private {!../../../src/service/url-replacements-impl.UrlReplacements} */
+    this.urlReplacements_ = Services.urlReplacementsForDoc(ampdoc);
+
+    /** @private @const {!../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = Services.viewerForDoc(ampdoc);
+
+    /** @private @const {?../../../src/service/performance-impl.Performance} */
+    this.performance_ = Services.performanceForOrNull(ampdoc.win);
+
+    /** @private @const {function(string):Promise<string>} */
+    this.openLoginDialog_ = openLoginDialog.bind(null, ampdoc);
+
+    /** @private {?JsonObject} */
+    this.authResponse_ = null;
+
+    /** @const @private {!SignInProtocol} */
+    this.signIn_ = new SignInProtocol(ampdoc, this.viewer_, this.pubOrigin_,
+      configJson);
+
+    /** @private {?Function} */
+    this.firstAuthorizationResolver_ = null;
+
+    /**
+     * Create a resolver we can call whenever we want. Attach analytics
+     * and use to block other behavior on our first and/or most recent
+     * auth call until resolver is explicitly called.
+     * @const @private {!Promise}
+     */
+    this.firstAuthorizationPromise_ = new Promise(resolve => {
+      this.firstAuthorizationResolver_ = resolve;
+    });
+
+    /** @private {!Object<string, string>} */
+    this.loginUrlMap_ = {};
+
+    /** @private {?Promise} */
+    this.loginPromise_ = null;
+
+    /** @private {!time} */
+    this.loginStartTime_ = 0;
+
+    this.firstAuthorizationPromise_.then(() => {
+      this.analyticsEvent_('access-authorization-received');
+      if (this.performance_) {
```

Comment:
Would seem like this should still belong in the top-level service.

Reply:
are you thinking this is still called once per source or just once globally?

Reply:
Once globally is enough.

Reply:
And better :)